### PR TITLE
Fix attribute owner

### PR DESCRIPTION
**Pull request by denis-ivanov**
# Types of Changes

## Prerequisites

Please make sure you can check the following two boxes:

- [x] I have read the **CONTRIBUTING** document
- [x] My code follows the code style of this project

## Contribution Type

What types of changes does your code introduce? Put an `x` in all the boxes that apply:

- [x] Bug fix (non-breaking change which fixes an issue, please reference the issue id)
- [ ] New feature (non-breaking change which adds functionality, make sure to open an associated issue first)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed

## Description

Pass attribute owner to the constructor+ tests.

### CODE REVIEW

File: src/AngleSharp.Core.Tests/Html/DOM.cs [hunk 1]
```diff
@@ -735,5 +735,15 @@ public void IteratingThroughDomShouldNeverFail_Issue763()
...
         }
+
+        [Test]
+        public void OwnerElement_ShouldNotBeNull()
+        {
+            var document = "<body><div class='one'></div></body>".ToHtmlDocument();
+            var div = document.QuerySelector("div");
+            var attr = div.Attributes["class"];
+
+            Assert.NotNull(attr.OwnerElement);
```
Comment (campersau) [inline src/AngleSharp.Core.Tests/Html/DOM.cs:746]: Would an additional `Assert.AreSame`
```suggestion
            Assert.NotNull(attr.OwnerElement);
            Assert.AreSame(div, attr.OwnerElement);
```
 make sense here?

File: src/AngleSharp/Dom/Internal/Attr.cs [hunk 2]
```diff
@@ -15,6 +15,7 @@ public sealed class Attr : IAttr
         private readonly String _localName;
         private readonly String? _prefix;
         private readonly String? _namespace;
+        private readonly IElement? _ownerElement;
```
Comment (FlorianRappl) [inline src/AngleSharp/Dom/Internal/Attr.cs]: Well, this adds 1 field per attribute. As there are most likely more attributes than elements the memory consumption should not be disregarded here.

We already can infer the owner element via the container. Is there a reason to duplicate this information?

**Review by FlorianRappl — COMMENTED** [review]
Well, in general is fine, but we should do something about the owner element field duplication. I'd rather like to see this only once in there. Maybe instead of passing the owner via the constructor we should pass the container? Or remove the owner from the container (but the container may need this info, too)?

File: src/AngleSharp/Html/Parser/HtmlDomBuilderExtensions.cs [hunk 3]
```diff
@@ -19,7 +19,7 @@ public static void SetAttributes(this Element element, List<HtmlAttributeToken>
             for (var i = 0; i < attributes.Count; i++)
             {
                 var attribute = attributes[i];
-                var item = new Attr(attribute.Name, attribute.Value);
+                var item = new Attr(attribute.Name, attribute.Value, element);
```
Comment (campersau) [inline src/AngleSharp/Html/Parser/HtmlDomBuilderExtensions.cs]:
```suggestion
                var item = new Attr(attribute.Name, attribute.Value);
                item.Container = container;
```

File: src/AngleSharp/Dom/Internal/Element.cs [hunk 4]
```diff
@@ -630,7 +630,7 @@ protected void CloneElement(Element element, Document owner, Boolean deep)
 
             foreach (var attribute in _attributes)
             {
-                var attr = new Attr(attribute.Prefix, attribute.LocalName, attribute.Value, attribute.NamespaceUri);
+                var attr = new Attr(attribute.Prefix, attribute.LocalName, attribute.Value, attribute.NamespaceUri, this);
```
Comment (campersau) [inline src/AngleSharp/Dom/Internal/Element.cs]:
```suggestion
                var attr = new Attr(attribute.Prefix, attribute.LocalName, attribute.Value, attribute.NamespaceUri);
                attr.Container = element._attributes;
```

**Review by campersau — COMMENTED** [review]
Would it work if the container is set manually everywhere `FastAddItem` is used?

E.g. in `Element.cs`:
```c#
        public void AddAttribute(Attr attr)
        {
            attr.Container = _attributes;
            _attributes.FastAddItem(attr);
        }
```

**Comment by FlorianRappl** [comment]
> Would it work if the container is set manually everywhere FastAddItem is used?

I think it would.

File: src/AngleSharp.Core.Tests/Html/DOM.cs (hunk 1 above)
Comment (denis-ivanov) [reply src/AngleSharp.Core.Tests/Html/DOM.cs:746]: @campersau, done.

File: src/AngleSharp/Dom/Internal/Attr.cs (hunk 2 above)
Comment (denis-ivanov) [reply src/AngleSharp/Dom/Internal/Attr.cs]: @FlorianRappl, it was a bad idea. I've removed this field.

File: src/AngleSharp/Html/Parser/HtmlDomBuilderExtensions.cs (hunk 3 above)
Comment (denis-ivanov) [reply src/AngleSharp/Html/Parser/HtmlDomBuilderExtensions.cs]: Fixed

File: src/AngleSharp/Dom/Internal/Element.cs (hunk 4 above)
Comment (denis-ivanov) [reply src/AngleSharp/Dom/Internal/Element.cs]: Fixed

File: src/AngleSharp/Dom/Internal/Attr.cs (hunk 2 above)
Comment (FlorianRappl) [reply src/AngleSharp/Dom/Internal/Attr.cs]: No worries - thanks for the awesome PR @denis-ivanov !

**Review by FlorianRappl — APPROVED** [review]
LGTM! Thanks :beers:

**Pull request merged**